src/xenia/cpu/stack_walker_posix.cc

```diff
@@ -9,16 +9,293 @@
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+#include <signal.h>
+
+#include "stack_walker.h"
 #include "xenia/base/logging.h"
+#include "xenia/cpu/backend/code_cache.h"
 
 namespace xe {
 namespace cpu {
 
+const int CAPTURE_SIGNAL = SIGUSR1;
+
+struct PosixStackCapture {
+  unw_context_t context_;
+  std::vector<unw_cursor_t> cursors_;
+
+  PosixStackCapture(unw_context_t&& context, size_t frame_offset,
+                    size_t frame_count)
+      : context_(context) {
+    unw_cursor_t cursor;
+    unw_init_local(&cursor, &context);
+
+    for (size_t i = 0; i < frame_offset; ++i) {
+      if (unw_step(&cursor) < 0) {
+        return;
+      }
+    }
+    for (uint32_t i = 0; i < frame_count; ++i) {
+      int step_result = unw_step(&cursor);
+      if (step_result == 0) {
+        break;
+      } else if (step_result < 0) {
+        switch (-step_result) {
+          case UNW_EUNSPEC:
+            return;
+          case UNW_ENOINFO:
+            return;
+          case UNW_EBADVERSION:
+            return;
+          case UNW_EINVALIDIP:
+            return;
+          case UNW_EBADFRAME:
+            return;
+          case UNW_ESTOPUNWIND:
+            return;
+          default:
+            return;
+        }
+      }
+      cursors_.push_back(cursor);
+    }
+  }
+};
+}  // namespace cpu
+}  // namespace xe
+
+namespace std {
+// TODO(bwrsandman): This needs to be expanded on to avoid collisions of stacks
+//  with same ip and depth
+template <>
+struct hash<xe::cpu::PosixStackCapture> {
+  typedef xe::cpu::PosixStackCapture argument_t;
+  typedef std::size_t result_t;
+  result_t operator()(argument_t const& capture) const noexcept {
+    if (capture.cursors_.empty()) {
+      return 0;
+    }
+    result_t result = 0;
+    unw_word_t ip;
+    unw_cursor_t front_cursor = capture.cursors_.front();
+    unw_get_reg(&front_cursor, UNW_REG_IP, &ip);
+    auto h1 = std::hash<unw_word_t>{}(ip);
+    auto h2 = std::hash<size_t>{}(capture.cursors_.size());
+    return h1 ^ (h2 << 1);
+  }
+};
+
+template <>
+struct equal_to<xe::cpu::PosixStackCapture> {
+  typedef xe::cpu::PosixStackCapture argument_t;
+  typedef std::size_t result_t;
+  result_t operator()(argument_t const& capture_1,
+                      argument_t const& capture_2) const noexcept {
+    if (capture_1.cursors_.size() != capture_2.cursors_.size()) {
+      return false;
+    }
+    unw_word_t reg_1, reg_2;
+    unw_cursor_t front_cursor_1 = capture_1.cursors_.front();
+    unw_cursor_t front_cursor_2 = capture_2.cursors_.front();
+    for (uint32_t i = 0; i < UNW_REG_LAST; ++i) {
+      unw_get_reg(&front_cursor_1, i, &reg_1);
+      unw_get_reg(&front_cursor_2, i, &reg_2);
+      if (reg_1 != reg_2) {
+        return false;
+      }
+    }
+    return true;
+  }
+};
+}  // namespace std
+
+namespace xe {
+namespace cpu {
+
+#include <libiberty/demangle.h>
+#include <memory>
+std::string demangle(const char* mangled_name) {
+  if (mangled_name == std::string()) {
```

Comment:
For checking if an `std::string` is empty, `.empty()` can be used. In this case (a C string), `[0]` can be compared to 0.